compiler/luci-interpreter/src/kernels/Tanh.cpp

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2020 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "kernels/Tanh.h"
+
+#include "kernels/Utils.h"
+
+#include <tensorflow/lite/kernels/internal/reference/reference_ops.h>
+
+namespace luci_interpreter
+{
+namespace kernels
+{
+
+Tanh::Tanh(const Tensor *input, Tensor *output) : Kernel({input}, {output}) {}
+
+void Tanh::configure()
+{
+  assert(input()->element_type() == output()->element_type());
+  if (input()->element_type() == DataType::U8)
+  {
+    assert(output()->scale() == 1. / 128);
+  }
+  output()->resize(input()->shape());
+}
+
+void Tanh::execute() const
+{
+  switch (input()->element_type())
+  {
+    case DataType::FLOAT32:
+      evalFloat();
+      break;
+    case DataType::U8:
+      evalQuantized();
+      break;
+    default:
+      throw std::runtime_error("Unsupported type.");
+  }
+}
+
+void Tanh::evalFloat() const
+{
+  tflite::reference_ops::Tanh(getTensorShape(input()), getTensorData<float>(input()),
+                              getTensorShape(output()), getTensorData<float>(output()));
+}
+
+void Tanh::evalQuantized() const
+{
+  const int size = tflite::MatchingFlatSize(getTensorShape(input()), getTensorShape(output()));
+  uint8_t *output_data = getTensorData<uint8_t>(output());
+  const uint8_t *input_data = getTensorData<uint8_t>(input());
+
+  tflite::reference_ops::Tanh(getTensorShape(input()), getTensorData<float>(input()),
+                              getTensorShape(output()), getTensorData<float>(output()));
```

Comment:
This is floating point execution. I think you should go back to your previous approach that uses a lookup table. Can you roll back the code to the point when you make a lookup table? I'll see what is the error.

Reply:
Thank you for pointing it out.
I dropped my latest 2 commits with rebase.

Thank you for your advice, @jinevening ! :)

Reply:
When I make temporary c++ file below:
<details><summary>Temp C++ Code</summary>
<pre><code>
#include <iostream>
#include <cmath>

using namespace std;

int main() {
int a[] = {0,  -6, 2, 4, -4, -2, 8, 1 };
float b[] = {0.0, -0.999987, 0.964027, 0.999329, -0.999329, -0.96402,  0.99999,  0.76159};
float scale = 1 / 128.f;
int zero_point = 0;


for(int i = 0; i < 8; i++) {
    cout << "-----------------------------------" << endl;
    cout << i << "th round" << endl;
    const float transformed = std::tanh(a[i]);
    cout << "Tanh value of input: " << transformed << endl;
    cout << "Expected output: " << b[i] << endl;
}

return 0;
}
</code></pre></details>

I get results below:
<details><summary>Result</summary>
<pre><code>
-----------------------------------
0th round
Tanh value of input: 0
Expected output: 0
-----------------------------------
1th round
Tanh value of input: -0.999988
Expected output: -0.999987
-----------------------------------
2th round
Tanh value of input: 0.964028
Expected output: 0.964027
-----------------------------------
3th round
Tanh value of input: 0.999329
Expected output: 0.999329
-----------------------------------
4th round
Tanh value of input: -0.999329
Expected output: -0.999329
-----------------------------------
5th round
Tanh value of input: -0.964028
Expected output: -0.96402
-----------------------------------
6th round
Tanh value of input: 1
Expected output: 0.99999
-----------------------------------
7th round
Tanh value of input: 0.761594
Expected output: 0.76159
</code></pre></details>

So, reference input & output from [Tensorflow Reference](https://github.com/tensorflow/tensorflow/blob/v2.1.0/tensorflow/lite/kernels/activations_test.cc#L615-L670) is quite correct (which is obvious),

But, what I get for my result of `Tanh.test.cpp` is like below:
<pre><code>
Actual: { 0, -0.757812, 0.757812, 0.757812, -0.757812, -0.757812, 0.757812, 0.757812, 0, -0.757812, 0.757812, 0.757812, -0.757812, -0.757812, 0.757812, 0.757812, 0, -0.757812, 0.757812, 0.757812, -0.757812, -0.757812, 0.757812, 0.757812, 0, -0.757812, 0.757812, 0.757812, -0.757812, -0.757812, 0.757812, 0.757812, ... }
</code></pre>

I don't get where these results are coming from.
Even inputs are not like {0, -4, 4, 4, -4, -4, ...}, but 0.757812 is keep coming out as a result.

Reply:
@jinevening ,

I changed my `tanh.test.cpp` code to below:
<details><summary><b>tanh.test.cpp</b></summary>
<pre><code>
TEST(TanhTest, Uint8)
{
  float kMin = -6;
  float kMax = 8;
  float kTanhTolerance = 2 * (1. / 256);
  std::pair<float, int32_t> quant_param = quantizationParams<uint8_t>(kMin, kMax);
  std::vector<float> input_data{
      0,  -6, 2, 4, //
      -4, -2, 8, 1, //
      0,  -6, 2, 4, //
      -4, -2, 8, 1, //
      0,  -6, 2, 4, //
      -4, -2, 8, 1, //
      0,  -6, 2, 4, //
      -4, -2, 8, 1, //
      0,  -6, 2, 4, //
      -4, -2, 8, 1, //
      0,  -6, 2, 4, //
      -4, -2, 8, 1, //
  };
  Tensor input_tensor{
      DataType::U8, {2, 6, 4, 1}, {{quant_param.first}, {quant_param.second}}, ""};
  Tensor output_tensor = makeOutputTensor(DataType::U8, quant_param.first, quant_param.second);
  std::vector<uint8_t> quantize_input =
      quantize<uint8_t>(input_data, quant_param.first, quant_param.second);
  input_tensor.writeData(quantize_input.data(), quantize_input.size() * sizeof(uint8_t));

  Tanh kernel(&input_tensor, &output_tensor);
  kernel.configure();
  kernel.execute();

  std::vector<float> output_data{
      0.0,       -0.999987, 0.964027, 0.999329, //
      -0.999329, -0.96402,  0.99999,  0.76159,  //
      0.0,       -0.999987, 0.964027, 0.999329, //
      -0.999329, -0.96402,  0.99999,  0.76159,  //
      0.0,       -0.999987, 0.964027, 0.999329, //
      -0.999329, -0.96402,  0.99999,  0.76159,  //
      0.0,       -0.999987, 0.964027, 0.999329, //
      -0.999329, -0.96402,  0.99999,  0.76159,  //
      0.0,       -0.999987, 0.964027, 0.999329, //
      -0.999329, -0.96402,  0.99999,  0.76159,  //
      0.0,       -0.999987, 0.964027, 0.999329, //
      -0.999329, -0.96402,  0.99999,  0.76159,  //
  };
  std::vector<int32_t> ref_output_shape{2, 6, 4, 1};
  EXPECT_THAT(dequantize<uint8_t>(extractTensorData<uint8_t>(output_tensor), output_tensor.scale(),
                                  output_tensor.zero_point()),
              ElementsAreArray(ArrayFloatNear(output_data, 8 * kTanhTolerance)));
  EXPECT_THAT(extractTensorShape(output_tensor), ::testing::ElementsAreArray(ref_output_shape));
}
</code></pre></details>

I only changed `kMin` to -6, and `kMax` to 8 (which is the provided min & max from input_data)
This gives output of below:
<pre><code>
Actual: { 0, -0.988235, 0.988235, 0.988235, -0.988235, -0.988235, 0.988235, 0.768627, 0, -0.988235, 0.988235, 0.988235, -0.988235, -0.988235, 0.988235, 0.768627, 0, -0.988235, 0.988235, 0.988235, -0.988235, -0.988235, 0.988235, 0.768627, 0, -0.988235, 0.988235, 0.988235, -0.988235, -0.988235, 0.988235, 0.768627, ... }
</code></pre>

This looks more like what it should be, because the actual `ref_output_data` looks like below:
<pre><code>
std::vector<float> output_data{
      0.0,       -0.999987, 0.964027, 0.999329, //
      -0.999329, -0.96402,  0.99999,  0.76159,  //
      0.0,       -0.999987, 0.964027, 0.999329, //
      -0.999329, -0.96402,  0.99999,  0.76159,  //
      0.0,       -0.999987, 0.964027, 0.999329, //
      -0.999329, -0.96402,  0.99999,  0.76159,  //
      0.0,       -0.999987, 0.964027, 0.999329, //
      -0.999329, -0.96402,  0.99999,  0.76159,  //
      0.0,       -0.999987, 0.964027, 0.999329, //
      -0.999329, -0.96402,  0.99999,  0.76159,  //
      0.0,       -0.999987, 0.964027, 0.999329, //
      -0.999329, -0.96402,  0.99999,  0.76159,  //
  };
</pre></code>

Seems like by putting `kMin = -1` * `kMax = 127. / 128.` into `quantizationParams`(which is referenced from Tensorflow Reference) limits the quantized input's maximum to 0.992188, and input's minimum to -1.
When I print a log for quantized_inputs, I get something like this:
<pre><code>
Quantized input_tensor before tanh kernel =
Actual: { 0, -1, 0.992188, 0.992188, -1, -1, 0.992188, 0.992188, 0, -1, 0.992188, 0.992188, -1, -1, 0.992188, 0.992188, 0, -1, 0.992188, 0.992188, -1, -1, 0.992188, 0.992188, 0, -1, 0.992188, 0.992188, -1, -1, 0.992188, 0.992188, ... }
</code></pre>

which looks like corrupted data.

As a result,
input data of -6 becomes -1 (kMin),
input data of 2 becomes 0.992188 (kMax),
input data of 4 becomes 0.992188 (kMax),
input data of -4  becomes -1 (kMin),
input data of -2 becomes -1 (kMin), and so on...

So all inputs becomes either kMin, kMax or 0.
Therefore Tanh calculated outputs also becomes either -0.757812, 0.757812, or 0.